src/client/csfle/state_machine.rs

```diff
@@ -0,0 +1,171 @@
+        // thread for crypto finalization.  An `Option` would also work here, but `Result` means we
+        // can return a helpful error if things get into a broken state rather than panicing.
+        let mut ctx = Ok(ctx);
+        loop {
+            let state = result_ref(&ctx)?.state()?;
+            match state {
+                State::NeedMongoCollinfo => {
+                    let ctx = result_mut(&mut ctx)?;
+                    let filter = raw_to_doc(ctx.mongo_op()?)?;
+                    let metadata_client = csfle
+                        .aux_clients
+                        .metadata_client
+                        .as_ref()
+                        .and_then(|w| w.upgrade())
+                        .ok_or_else(|| {
+                            Error::internal("metadata_client required for NeedMongoCollinfo state")
+                        })?;
+                    let db = metadata_client.database(db.as_ref().ok_or_else(|| {
+                        Error::internal("db required for NeedMongoCollinfo state")
+                    })?);
+                    let mut cursor = db.list_collections(filter, None).await?;
+                    if cursor.advance().await? {
+                        ctx.mongo_feed(cursor.current())?;
+                    }
+                    ctx.mongo_done()?;
+                }
+                State::NeedMongoMarkings => {
+                    let ctx = result_mut(&mut ctx)?;
+                    let command = ctx.mongo_op()?.to_raw_document_buf();
+                    let db = db.as_ref().ok_or_else(|| {
+                        Error::internal("db required for NeedMongoMarkings state")
+                    })?;
+                    let op = RawOutput(RunCommand::new_raw(db.to_string(), command, None, None)?);
+                    let mongocryptd_client = csfle
+                        .mongocryptd_client
+                        .as_ref()
+                        .ok_or_else(|| Error::internal("mongocryptd client not found"))?;
+                    let response = mongocryptd_client.execute_operation(op, None).await?;
+                    ctx.mongo_feed(response.raw_body())?;
+                    ctx.mongo_done()?;
+                }
+                State::NeedMongoKeys => {
+                    let ctx = result_mut(&mut ctx)?;
+                    let filter = raw_to_doc(ctx.mongo_op()?)?;
+                    let kv_ns = &csfle.opts.key_vault_namespace;
+                    let kv_client = csfle
+                        .aux_clients
+                        .key_vault_client
+                        .upgrade()
+                        .ok_or_else(|| Error::internal("key vault client dropped"))?;
+                    let kv_coll = kv_client
+                        .database(&kv_ns.db)
+                        .collection::<RawDocumentBuf>(&kv_ns.coll);
+                    let mut cursor = kv_coll.find(filter, None).await?;
+                    while let Some(result) = cursor.try_next().await? {
```

Comment:
It's a little bit faster to use the `cursor.advance` and `cursor.current` API here to avoid having to trip through `serde`.